Composable-transactionMint

daiagi opened the pull request:
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [ ] Feature
- [x] Refactoring

## Context

- [x] Closes #5034
- [ ] Requires deployment <>

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [x] tested by minting on snek and RMRK, with and without listing for sell
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=EfmnRhHaQqfT3phm4cUCHCU3gFVDoSPR1U9WXzMRQBMqZ4L)

#### Community participation

- [ ] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

daiagi commented:
this PR covers minting Token composable
Minting collection will follow in another PR

vikiival commented on composables/transaction/transactionMintToken.ts:
Sorry to be harsh but this is bit useless to take the same code from two different files to put it under each other

daiagi replied:
:disappointed: 
don't know what to tell you
they have different logic...
it's not as if the only thing i did was copy paste and put it together in one file

vikiival reviewed (changes requested):
^

vikiival commented:
> Minting collection will follow in another PR


I would rather do this first

vikiival commented:
Yeah sorry, closing that was a bit harsh. Sorry for that

I just rather do the collections first, it is much easier and 80% of pinning logic can be extracted away. 

The reasoning is this is hard to review. Contains a lot of dupictated code.

daiagi commented:
Yeah, it was challenging to write as well

I'll do the collection minting, sure

But as for this one, how about I split the files to make it easier read? 

And there are actually no duplications
In fact I've removed duplicate code

vikiival commented:
> But as for this one, how about I split the files to make it easier read?

Yes please decompose pinning from the file 

> I'll do the collection minting, sure

Yeah I think it's easier to do.

Sorry one more time for closing that 🙈

vikiival commented on composables/transaction/MintToken/transactionMinRmrk.ts:
You have typo in the file name

daiagi replied:
yy. fix in a second. I rushed

roiLeo commented:
Hey! I don't want to interfere but I've just tested on snek (`/snek/gallery/331660682-9`) and for me it works very well.
It's already a good start and it allows to get the logic out of the components (better readability). And that is what I asked when I've open #5034.
In any case it will be useful when we'll rewrite minting experence (and maybe massmint).
Nice work, keep it up! 😉

daiagi commented:
> Hey! I don't want to interfere but I've just tested on snek (`/snek/gallery/331660682-9`) and for me it works very well. It's already a good start and it allows to get the logic out of the components (better readability). And that is what I asked when I've open #5034. In any case it will be useful when we'll rewrite minting experence (and maybe massmint). Nice work, keep it up! wink


Thanks @roiLeo 
Appreciate the positive feedback :smile:

vikiival commented:
Wondering why are you refactoring SimpleMint instead of rmrk/CreateToken.vue ? 

<img width="347" alt="Screenshot 2023-03-15 at 14 48 16" src="https://user-images.githubusercontent.com/22471030/225328670-0d892bdc-1962-4420-8cec-61b2e72bdf7c.png">

vikiival commented:
But so far really good job 🤗

roiLeo commented on utils/mintUtils.ts:
```diff
@@ -1,6 +1,6 @@
-import { MassMintNFT } from '../service/scheme'
-import { MediaType } from '../types'
-import { resolveMedia } from '../utils'
+import { MassMintNFT } from '../components/rmrk/service/scheme'
+import { MediaType } from '../components/rmrk/types'
+import { resolveMedia } from '../components/rmrk/utils'
```
```suggestion
import { MassMintNFT } from '@/components/rmrk/service/scheme'
import { MediaType } from '@/components/rmrk/types'
import { resolveMedia } from '@/components/rmrk/utils'
```

IMO some of this files should be put outside rmrk folder (maybe /components/base) but it's not on this scope

roiLeo reviewed (approved):
otherwise wfm ✅

vikiival commented:
![image](https://user-images.githubusercontent.com/22471030/225584217-985275f5-0320-4c55-adb6-ed60483f3ae7.jpeg)

vikiival commented:
Looks like some failed checks

daiagi commented:
> Looks like some failed checks

@vikiival they can be ignored because
1.
![image](https://user-images.githubusercontent.com/22791238/225604613-13e59b32-244a-4bdc-ab3f-90c7572232bc.png)

both these function will return for chains bsx, snek, rmrk, rmrk2, as needed, DeepSource complains that there isn't a return value outside of an if, and there shouldn't be

2.
![image](https://user-images.githubusercontent.com/22791238/225605076-e2b59dc7-2dbe-49d7-8cc3-1b81ce8f73ba.png)

this is about  the signature of ```executeTransaction``` \ ```howAboutToExecute```. I'm not gonna mess with it in this PR

roiLeo commented:
> Update: created composables for minting collection and use them in CreateCollection components
> 
> @roiLeo have a look once again please

Now same issue as #5266 (collectionId: `3124270296`) 

![Screenshot 2023-03-16 at 13-21-58 KodaDot Low fees and low carbon minting KodaDot - NFT Market Explorer](https://user-images.githubusercontent.com/9987732/225616038-cb9229b5-4e54-490e-bbc2-8a54b49f8f55.png)

I saw somewhere that `[[ Rare Waifu ]]` was default fallback metadata

daiagi commented:
> Now same issue as #5266 (collectionId: `3124270296`)

What chain?
I'll double check.

daiagi commented:
ok so...
1. it happens when minting collection on snek in beta as well
2. I've tested the minting collection in this PR in RMRK and BSX, both works

so the issue does not stem from this PR

vikiival commented on components/bsx/Create/CreateToken.vue:
```diff
@@ -287,47 +274,45 @@ export default class CreateToken extends mixins(
+        tags: this.attributes,
+        edition,
+        successMessage: (blockNumber) =>
+          `NFT ${name} Saved in block ${blockNumber}`,
```
Maybe put into translations?

daiagi replied:
Done

vikiival commented on components/rmrk/Create/CreateCollection.vue:
```diff
@@ -189,43 +131,37 @@ export default class CreateCollection extends mixins(
+      transaction({
+        interaction: Interaction.MINT,
+        urlPrefix: usePrefix().urlPrefix.value,
+        description,
+        file,
+        name,
+        tags: [],
+        count: this.unlimited ? 0 : this.max,
+        symbol: this.symbol,
```
Isn't better to wrap it into the other payload object?

daiagi replied:
agree
I've changed the signature to 
```ts
export interface ActionMintCollection {
  interaction: Interaction.MINT
  urlPrefix: string
  collection: CollectionToMintBasilisk | CollectionToMintKusama
  successMessage?: string | ((blockNumber: string) => string)
  errorMessage?: string
}
```

vikiival commented on composables/transaction/mintCollection/constructMeta.ts:
```diff
@@ -0,0 +1,57 @@
+  tags: Attribute[]
+  urlPrefix: string
+}) => {
+  if (urlPrefix === 'bsx' || urlPrefix === 'snek') {
+    return tags.map((val) => ({
+      ...val,
+      display_type: null,
+    }))
+  }
+  if (urlPrefix === 'rmrk' || urlPrefix === 'rmrk2') {
+    return []
+  }
```
why so? tags should be also available in rmrk

daiagi replied:
see line 166 in /rmrk/create/createCollection (current code, in main)

createMetaData for RMRK collection was passed empty array for attributes

vikiival commented on composables/transaction/mintToken/transactionMintRmrk.ts:
```diff
@@ -0,0 +1,82 @@
+import { constructMeta } from './constructMeta'
+import { ActionMintToken } from '../types'
+
+type RMRKMintedCollection = BaseMintedCollection & {
```
```suggestion
type KusamaMintedCollection = BaseMintedCollection & {
```

daiagi replied:
renamed and split:

```ts
export type MintedCollectionKusama = BaseMintedCollection & {
  max: number
  symbol: string
}
export type MintedCollectionBasilisk = BaseMintedCollection & {
  lastIndexUsed: number
}
```

vikiival commented on utils/mintUtils.ts:
```diff
@@ -73,8 +73,8 @@ function toMassMint(mints: string[][]) {
 export const isRangeSyntax = (text: string): boolean => {
-  const r = /^\d+-\d*\n/
-  return r.test(text)
+  const regExp = /^\d+-\d*\n/
+  return regExp.test(text)
```
```suggestion
  const regex = /^\d+-\d*\n/
  return regex.test(text)
```

vikiival commented on utils/mintUtils.ts:
```diff
@@ -89,19 +89,19 @@ export function isMatchAll(text: string): boolean {
 }
 
+const hasIndex = (line: string) => {
+  const regExp = /{i}/
+  return regExp.test(line)
```
```suggestion
  const regex = /{i}/
  return regex.test(line)
```

vikiival commented on utils/mintUtils.ts:
```diff
@@ -89,19 +89,19 @@ export function isMatchAll(text: string): boolean {
 export const skipProcess = (line: string): boolean => {
-  const r = /^-/
-  return r.test(line)
+  const regExp = /^-/
+  return regExp.test(line)
```
```suggestion
  const regex = /^-/
  return regex.test(line)
```

vikiival commented on utils/mintUtils.ts:
```diff
@@ -113,8 +113,8 @@ const correctText = (original: string, parsed: string): string => {
 export function toRange(line: string): Range | null {
-  const r = /^(\d+)-(\d*)\n?$/
-  const match = r.exec(line)
+  const regExp = /^(\d+)-(\d*)\n?$/
+  const match = regExp.exec(line)
```
```suggestion
  const regex = /^(\d+)-(\d*)\n?$/
  const match = regex.exec(line)
```

helloitsdamsky commented:
It currently doesn’t let me mint with KSM fee asset. Says "Insufficient funds"  even tho i have enough :)

daiagi commented:
> It currently doesn’t let me mint with KSM fee asset. Says "Insufficient funds"  even tho i have enough :)

Hey @helloitsdamsky
Do I understand correctly that you tried on kusama network?
Did you try minting a collection or NFT?
does it work for you on beta?

Thank you

daiagi commented on components/rmrk/Create/CreateCollection.vue:
```diff
@@ -136,46 +113,6 @@ export default class CreateCollection extends mixins(
-      description,
-      imageHash,
-      undefined,
-      [],
```
@vikiival here

daiagi commented:
tested again by minting collection and NFT on BSX and RMRK
* /bsx/collection/651189115
* /rmrk/collection/669DB118F2897C6014-PLN

![image](https://user-images.githubusercontent.com/22791238/226088221-0260c75d-1ca9-4e71-93ba-ad52761145fe.png)
![image](https://user-images.githubusercontent.com/22791238/226088250-f73112a5-9002-401f-93f2-d2ee5dec816b.png)

vikiival commented:
Can please @JustLuuuu or @helloitsdamsky test?

daiagi commented:
/reviewpad summerize

daiagi commented:
/reviewpad summarize

daiagi commented:
> I have created a new wallet especially to test this while I was writing an article about minting on Basilisk. This account has exactly 0 bsx tokens. Only KSM (bridged). I believe even if you set KSM as a fee, it still somehow requires bsx. Test it with a wallet where you have 0 bsx.
> 
> When I tried to test it with a wallet where I had some bsx tokens. It worked.

strongly suspect it has to do with this:
* https://github.com/kodadot/nft-gallery/issues/5284

daiagi commented:
I had a chat with @JustLuuuu  who confirmed minting on Kusama works
I have also created PR #5305 to fix the minting with bridged KSM on Basilisk issue

vikiival commented:
More like suggestions otherwise LGTM

daiagi commented on composables/useTransaction.ts:
```diff
@@ -9,44 +12,78 @@ import { execBurnTx } from './transaction/transactionBurn'
+  errorMessage?: string
+}
+
+const resolveSuccessMessage = (successMessage, block: string) => {
+  if (typeof successMessage === 'string') {
+    return successMessage
+  }
+  if (typeof successMessage === 'function') {
+    return successMessage(block)
+  }
+
+  return 'Success!'
+}
```
sure why not

daiagi replied:
done

daiagi commented on components/rmrk/Create/CreateToken.vue:
```diff
@@ -238,111 +212,52 @@ export default class CreateToken extends mixins(
+      watch([blockNumber, createdNFTs], () => {
+        if (this.hasPrice) {
+          if (blockNumber.value && createdNFTs.value) {
+            setTimeout(
+              () =>
+                this.listForSale(
+                  createdNFTs.value,
+                  blockNumber.value as string
+                ),
+              300
```
i was wondering the same, but no. it works

daiagi commented on components/bsx/Create/CreateToken.vue:
```diff
@@ -81,50 +81,33 @@
-  offsetAttribute,
-  secondaryFileVisible,
-} from '@/components/rmrk/Create/mintUtils'
+import { offsetAttribute } from '@/utils/mintUtils'
```
oh ok, this is not something I've changed in this PR
It's marked because I've moved mintUtils

remove it?

daiagi commented on composables/transaction/mintToken/transactionMintBasilisk.ts:
```diff
@@ -0,0 +1,60 @@
+
+  const create = api.tx.nft.mint(collectionId, nextId, metadata)
+
+  const list = price
```
sorry i don't follow
are you saying i should check that price isn't 0?

daiagi replied:
resolved 

```ts
  const list = Boolean(Number(price))
    ? [api.tx.marketplace.setPrice(collectionId, nextId, price)]
    : []
```

vikiival commented on composables/transaction/mintToken/transactionMintBasilisk.ts:
```diff
@@ -0,0 +1,60 @@
+
+  const create = api.tx.nft.mint(collectionId, nextId, metadata)
+
+  const list = Boolean(Number(price))
```
```suggestion
  const list = Number(price) > 0
```

daiagi replied:
Yeah? You prefer it that way?

yangwao commented:
😍 Perfect, I’ve sent the payout 
💵 $100 @ 35.66 USD/KSM ~ 2.804 $KSM
🧗 [EfmnRhHaQqfT3phm4cUCHCU3gFVDoSPR1U9WXzMRQBMqZ4L](https://kusama.subscan.io/account/EfmnRhHaQqfT3phm4cUCHCU3gFVDoSPR1U9WXzMRQBMqZ4L)
🔗 [0xc516fcda2389cb05b71352d71bcc3885de14deface541810ad216fbb1a4dc5ab](https://kusama.subscan.io/extrinsic/0xc516fcda2389cb05b71352d71bcc3885de14deface541810ad216fbb1a4dc5ab)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)
